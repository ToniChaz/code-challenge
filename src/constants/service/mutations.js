const CREATE = `{
  article {
    author
    excerpt
    content
    id
    title
    published
    tags
  }
}`

const UPDATE = `{
  article {
    author
    excerpt
    content
    id
    title
    published
    tags
  }
}`

const DELETE = `
  mutation Mutations($id: String!){
    deleteArticle(id: $id){
      id
    }
  }
`

export {
  CREATE,
  UPDATE,
  DELETE
}
