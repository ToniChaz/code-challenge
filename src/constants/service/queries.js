export const ARTICLES_QUERY = `{
  articles {
    author
    excerpt
    content
    id
    title
    published
    tags
  }
}`
