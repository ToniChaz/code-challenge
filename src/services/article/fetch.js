import { requestHandler } from '../serviceHandler'

export default function(query) {
  return requestHandler('/graphql', { query })
}
