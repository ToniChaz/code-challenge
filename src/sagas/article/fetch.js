import { call, put, takeEvery } from 'redux-saga/effects'
import articleService from '../../services/article/fetch'
import action from '../../actions/article'
import alertAction from '../../actions/alert'
import type from '../../constants/actions/article'
import { ARTICLES_QUERY } from '../../constants/service/queries'

export function* fetchArticles() {
  try {
    const response = yield call(articleService, ARTICLES_QUERY)
    yield put(action.fetchArticlesSuccess(response.data.articles))
  } catch (error) {
    yield put(alertAction.openAlert('danger', `An error occurred: ${error.message}`))
  }
}

export function* watchFetchArticles() {
  yield takeEvery(type.FETCH_ARTICLES, fetchArticles)
}
