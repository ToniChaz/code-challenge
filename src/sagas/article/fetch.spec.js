// fetch.spec.js

import { watchFetchArticles, fetchArticles } from './fetch'
import { ARTICLES_QUERY } from '../../constants/service/queries'
import { articles } from '../../../test/mocks'

describe('fetchArticles saga', () => {
  test('fetchArticles should trigger on every FETCH_ARTICLE action', () => {
    const generator = watchFetchArticles()

    expect(generator.next().value).toMatchSnapshot()
  })

  test('fetchArticles success', () => {
    const generator = fetchArticles(ARTICLES_QUERY)
    const response = {
      data: {
        articles
      }
    }

    expect(generator.next().value).toMatchSnapshot()
    expect(generator.next(response).value).toMatchSnapshot()
    expect(generator.next().done).toMatchSnapshot()
  })

  test('fetchArticles error', () => {
    const generator = fetchArticles(ARTICLES_QUERY)

    expect(generator.next().value).toMatchSnapshot()
    expect(generator.throw({ message: 'Error test' }).value).toMatchSnapshot()
    expect(generator.next().done).toMatchSnapshot()
  })
})
