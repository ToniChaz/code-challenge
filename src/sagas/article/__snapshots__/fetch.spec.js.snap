exports[`fetchArticles saga fetchArticles error 1`] = `
Object {
  "@@redux-saga/IO": true,
  "CALL": Object {
    "args": Array [
      "{
  articles {
    author
    excerpt
    content
    id
    title
    published
    tags
  }
}",
    ],
    "context": null,
    "fn": [Function],
  },
}
`;

exports[`fetchArticles saga fetchArticles error 2`] = `
Object {
  "@@redux-saga/IO": true,
  "PUT": Object {
    "action": Object {
      "color": "danger",
      "text": "An error occurred: Error test",
      "type": "OPEN_ALERT",
    },
    "channel": null,
  },
}
`;

exports[`fetchArticles saga fetchArticles error 3`] = `true`;

exports[`fetchArticles saga fetchArticles should trigger on every FETCH_ARTICLE action 1`] = `
Object {
  "@@redux-saga/IO": true,
  "FORK": Object {
    "args": Array [
      "FETCH_ARTICLES",
      [Function],
    ],
    "context": null,
    "fn": [Function],
  },
}
`;

exports[`fetchArticles saga fetchArticles success 1`] = `
Object {
  "@@redux-saga/IO": true,
  "CALL": Object {
    "args": Array [
      "{
  articles {
    author
    excerpt
    content
    id
    title
    published
    tags
  }
}",
    ],
    "context": null,
    "fn": [Function],
  },
}
`;

exports[`fetchArticles saga fetchArticles success 2`] = `
Object {
  "@@redux-saga/IO": true,
  "PUT": Object {
    "action": Object {
      "articles": Array [
        Object {
          "_id": "59eb8261d2ddc040ead6c7fa",
          "author": "Bertram Wintheiser",
          "content": "Et iure enim saepe. Repellat reprehenderit sequi id. Eveniet sit ex. Rem et labore temporibus dicta architecto reprehenderit qui. Consequatur amet aspernatur porro.",
          "excerpt": "Et iure enim saepe. Repellat reprehenderit sequi id. Eveniet sit ex. Rem et labore temporibus dicta ",
          "published": false,
          "tags": Array [
            "Avon portal Outdoors",
            "networks Berkshire Fish",
          ],
          "title": "Senior Communications Coordinator",
        },
        Object {
          "_id": "59eb8261d2ddc040ead6c7fb",
          "author": "Asha Koelpin DDS",
          "content": "Hic ut quia ipsam. Atque neque nam molestiae earum explicabo ab. Molestias soluta molestiae rem similique tenetur. Maiores doloribus rerum corrupti sed voluptatum nostrum delectus facere et.",
          "excerpt": "Hic ut quia ipsam. Atque neque nam molestiae earum explicabo ab. Molestias soluta molestiae rem simi",
          "published": true,
          "tags": Array [
            "New Mexico moderator Practical",
            "Ethiopian Birr",
          ],
          "title": "Internal Tactics Supervisor",
        },
      ],
      "type": "FETCH_ARTICLES_SUCCESS",
    },
    "channel": null,
  },
}
`;

exports[`fetchArticles saga fetchArticles success 3`] = `true`;
